documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt

```diff
@@ -0,0 +1,146 @@
+            Core.mixChannels(sources, destinations, fromTo)
+
+            // Try several threshold levels.
+            for (l in 0 until THRESHOLD_LEVEL) {
+                if (l == 0) {
+                    // HACK: Use Canny instead of zero threshold level.
+                    // Canny helps to catch squares with gradient shading.
+                    // NOTE: No kernel size parameters on Java API.
+                    Imgproc.Canny(gray0, gray, 10.0, 20.0)
+
+                    // Dilate Canny output to remove potential holes between edge segments.
+                    Imgproc.dilate(gray, gray, Mat.ones(Size(3.0, 3.0), 0))
+                } else {
+                    val threshold = (l + 1) * 255 / THRESHOLD_LEVEL
+                    Imgproc.threshold(
+                        gray0,
+                        gray,
+                        threshold.toDouble(),
+                        255.0,
+                        Imgproc.THRESH_BINARY
+                    )
+                }
+
+                // Find contours and store them all as a list.
+                Imgproc.findContours(
+                    gray,
+                    contours,
+                    Mat(),
+                    Imgproc.RETR_LIST,
+                    Imgproc.CHAIN_APPROX_SIMPLE
+                )
+                for (contour in contours) {
+                    val contourFloat = OpenCvUtils.toMatOfPointFloat(
+                        contour
+                    )
+                    val arcLen = Imgproc.arcLength(contourFloat, true) * 0.02
+
+                    // Approximate polygonal curves.
+                    val approx = MatOfPoint2f()
+                    Imgproc.approxPolyDP(contourFloat, approx, arcLen, true)
+                    if (isRectangle(approx, srcArea)) {
+                        rectangles.add(approx)
+                    }
+                }
+            }
+        }
+        return rectangles
+    }
+
+    private fun isRectangle(polygon: MatOfPoint2f, srcArea: Int): Boolean {
+        val polygonInt = OpenCvUtils.toMatOfPointInt(polygon)
+        if (polygon.rows() != 4) {
+            return false
+        }
+        val area = abs(Imgproc.contourArea(polygon))
+        if (area < srcArea * AREA_LOWER_THRESHOLD || area > srcArea * AREA_UPPER_THRESHOLD) {
+            return false
+        }
+        if (!Imgproc.isContourConvex(polygonInt)) {
+            return false
+        }
+
+        // Check if the all angles are more than 72.54 degrees (cos 0.3).
+        var maxCosine = 0.0
+        val approxPoints = polygon.toArray()
+        for (i in 2..4) {
+            val cosine = abs(
+                OpenCvUtils.angle(
+                    approxPoints[i % 4], approxPoints[i - 2], approxPoints[i - 1]
+                )
+            )
+            maxCosine = max(cosine, maxCosine)
+        }
+        return maxCosine < 0.3
+    }
+
+    private fun getAreaDescendingComparator(): Comparator<MatOfPoint2f> =
```

Comment:
Add method documentation.

Reply:
Methods on this class are related with the OpenCV algorithm inherited from the forked lib. It was created by the original author who separated the logic based on his own criteria.

Reply:
Could we have at least a minimum documentation explaining that?